text/0060-bazel.md

```diff
@@ -0,0 +1,137 @@
+---
+feature name: bazel
+start date: 2020-01-16
+rfc pr: (leave this empty)
+related issue: 0060
+---
+
+# Summary
+
+This proposal is for migrating the build system to [Bazel](https://bazel.build/), as it solves
+many if not all of the deficiencies of the current system. This involves adding Bazel resources and rules to the repository to manage
+build and test workflows, and bringing the team up to speed on how to use them.
+
+# Motivation
+
+The CDK currently uses a combination of [Lerna](https://lerna.js.org/), [Yarn Workspaces](https://yarnpkg.com/lang/en/docs/workspaces/), 
+and Bash scripts to orchestrate build and test workflows. This has several advantages for managing a monorepo structure, but can also
+be very slow, difficult to work with and extend, and hard to parallelize amongst the team (ie when one unrelated resource changes, the
+whole project still needs to be rebuilt). 
+
+# Basic Example
+
+Instead of running `bash build.sh` or `npm run build` on a subpackage, an engineer would simply run `bazel build <path-to-resource>` or
+`bazel test <path-to-resource`. Running a specific Bazel rule (for instance orchestrating a release) is also straightforward, with
+`bazel run <path-to-resource>.<action>`.
+
+# Design Summary
+
+In order to setup Bazel, the Bazel-specific rules and resources need to be added to the repository. This means adding `BUILD.bazel` files
+to each subpackage, and then building custom rules around some of the internals of the project (e.g. CFN to TS).
+
+# Detailed Design
+
+The structure of adding Bazel to a project is quite simple. Represented below is a tree of the files
+needed for a standard project
+
+```
+BAZEL.build
+package.json
+|packages|
+  BUILD.bazel
+  |apigateway|
+    BUILD.bazel
+    package.json
+    apigateway.ts
+    index.ts
+WORKSPACE <-- Bazel project definition, only needed once
+yarn.lock
+```
+
+The `BUILD` files themselves are quite simple, since Bazel has rules built-in for compiling a TypeScript library and for
+running JavaScript tests. We would need to write custom rules in the 
```

Comment:
The CDK is built with the `jsii` compiler (which uses typescript under the hood). What are the implications?

Also, what are the implications on all the tools in `tools/cdk-build-tools`?

Reply:
Each of those tools would get wrapped with Bazel rules. This would "supercharge" them with the benefits of Bazel with minimal added work to migrate the existing build system. It also gives a great template for adding more scripts to the `tools` part of the repository.

The implications for `jsii` are exactly the same, we would just wrap invoking `jsii` with a Bazel rule, which augments it with incrementality and all of the other Bazel-associated benefits.